_posts/2020-05-05-churn-vs-complexity-vs-coverage.markdown

```diff
@@ -0,0 +1,255 @@
+that the file has changed in the SCM that you are using)
+
+```
+**********************************************************************
+* Project Churn
+**********************************************************************
+
+Files
++-----------------------------------------------------------------------------------+---------------+
+| file_path                                                                         | times_changed |
++-----------------------------------------------------------------------------------+---------------+
+| lib/metric_fu.rb                                                                  | 50            |
+| lib/metric_fu/version.rb                                                          | 44            |
+| lib/metric_fu/configuration.rb                                                    | 36            |
+| lib/metric_fu/reporting/templates/awesome/awesome_template.rb                     | 29            |
+| lib/metric_fu/io.rb                                                               | 23            |
+| lib/metric_fu/formatter/html.rb                                                   | 23            |
+| lib/base/configuration.rb                                                         | 23            |
+| lib/metric_fu/run.rb                                                              | 22            |
+| lib/metric_fu/metrics/saikuro/saikuro.rb                                          | 20            |
+| lib/metric_fu/metric.rb                                                           | 20            |
++-----------------------------------------------------------------------------------+---------------+
+```
+
+By itself, this metric provides a limited insight:
+
+> These are the files that have changed the most since the beginning of the
+project
+
+It doesn't provide insights into why these files are changing as much as they
+do. If you open these files in your editor, you might find that files with high
+churn are super complex or super simple.
+
+<h2 id="complexity">Complexity</h2>
+
+As you go about making your changes, you will introduce complexity. There is no
+way around it. The best way to have no complexity is to have no code.
+
+It doesn't matter whether you are a rockstar object-oriented architect or a
+novice programmer. You will introduce complexity into a project.
+
+"Complexity" should not be considered a negative term. "Extreme complexity"
+might be considered an anti-pattern. Good object-oriented programming calls
+for [loose coupling and high cohesion](https://wiki.c2.com/?CouplingAndCohesion).
+
+You want modules that follow [SOLID](http://www.butunclebob.com/ArticleS.UncleBob.PrinciplesOfOod) 
+principles. When you notice your module is becoming extremely complex, you will
+want to refactor it into two or more modules. These modules will send and receive
+messages between each other, which will increase coupling and improve cohesion
```

Comment:
We want modules decrease coupling, it seems like this is implying that high coupled modules are a good thing?

Reply:
@bronzdoc the thing is that if you go from one module to n collaborating modules, you end up with more coupling. You go from (one module, no coupling), to (n modules, more coupling than no coupling)

I'll try to rephrase this.